src/main/scala/firrtl/passes/InferReadWrite.scala

```diff
@@ -0,0 +1,182 @@
+import firrtl.PrimOps._
+import Annotations._
+
+case class InferReadWriteAnnotation(t: String, tID: TransID = TransID(-1))
+    extends Annotation with Loose with Unstable {
+  val target = CircuitName(t)
+  def duplicate(n: Named) = this.copy(t=n.name)
+}
+
+object InferReadWritePass extends Pass {
+  def name = "Infer ReadWrite Ports"
+
+  def inferReadWrite(m: Module) = {
+    import WrappedExpression.we
+    val connects = HashMap[String, Expression]()
+    val repl = HashMap[String, Expression]()
+    val stmts = ArrayBuffer[Statement]()
+    val zero = we(UIntLiteral(0, IntWidth(1)))
+    val one = we(UIntLiteral(1, IntWidth(1)))
+
+    def analyze(s: Statement): Unit = s match {
+      case s: Connect  =>
+        connects(s.loc.serialize) = s.expr
+      case s: PartialConnect =>
+        connects(s.loc.serialize) = s.expr
+      case s: DefNode =>
+        connects(s.name) = s.value
+      case s: Block =>
+        s.stmts foreach analyze
+      case _ =>
+    }
+
+    def getProductTermsFromExp(e: Expression): Seq[Expression] =
+      e match {
+        // No ConstProp yet...
+        case Mux(cond, tval, fval, _) if we(tval) == one && we(fval) == zero =>
+          cond +: getProductTerms(cond.serialize)
+        case DoPrim(op, args, consts, tpe) if op == And =>
+          e +: (args flatMap getProductTermsFromExp)
+        case _: WRef | _: SubField | _: SubIndex | _: SubAccess =>
+          e +: getProductTerms(e.serialize)
+        case _ =>
+          List(e)
+      }
+
+    def getProductTerms(node: String): Seq[Expression] =
+      if (connects contains node) getProductTermsFromExp(connects(node)) else Nil
+
+    def checkComplement(a: Expression, b: Expression) = (a, b) match {
+      case (_, DoPrim(op, args, _, _)) if op == Not =>
+        args.head.serialize == a.serialize
+      case (DoPrim(op, args, _, _), _) if op == Not =>
+        args.head.serialize == b.serialize
+      case (_, DoPrim(op, args, _, _)) if op == Eq =>
+        args(0).serialize == a.serialize && we(args(1)) == zero ||
+        args(1).serialize == a.serialize && we(args(0)) == zero
+      case (DoPrim(op, args, _, _), _) if op == Eq =>
+        args(0).serialize == b.serialize && we(args(1)) == zero ||
+        args(1).serialize == b.serialize && we(args(0)) == zero
+      case _ => false
+    }
+
+    def inferReadWrite(s: Statement): Statement = s map inferReadWrite match {
+      case mem: DefMemory if mem.readLatency > 0 =>
+        var idx = 0
+        val bt = UIntType(IntWidth(1))
+        val ut = UnknownType
+        val ug = UNKNOWNGENDER
+        val readers = HashSet[String]()
+        val writers = HashSet[String]()
+        val readwriters = ArrayBuffer[String]()
+        for (w <- mem.writers ; r <- mem.readers) {
+          val wp = getProductTerms(s"${mem.name}.$w.en")
+          val rp = getProductTerms(s"${mem.name}.$r.en")
+          if (wp exists (a => rp exists (b => checkComplement(a, b)))) {
+            val rw = s"rw_$idx"
```

Comment:
Will conflict if there exists a readwriter of the name rw_0, you need to use the Uniquify class.